File: create-ts-bigconfig-package/templates/src/pkg/describe.ts
/** Simple package description. Pure report + workflow step. */
import { type Opts } from "big-config";
import { readBcPars } from "big-config/workflow";
import { okAlias, paramsOf, profileOf, toBcOpts } from "./interop.js";

export interface DescribeResult {
  profile: string;
  package: string;
  name: string;
}

function display(value: unknown, fallback = "(unset)"): string {
  return typeof value === "string" && value.trim() !== "" ? value : value == null ? fallback : String(value);
}

/** Pure: build a concise description of the active profile. */
export function describeReport(
  opts: Opts,
  env: Record<string, string | undefined> = process.env,
): DescribeResult {
  const merged = readBcPars(toBcOpts(opts), env);
  const params = paramsOf(merged);
  return {
    profile: display(profileOf(merged), "default"),
    package: display(params["package"]),
    name: display(params["name"]),
  };
}

export function printReport(result: DescribeResult): void {
  console.log(`Package: ${result.package}`);
  console.log(`Profile: ${result.profile}`);
  console.log(`Name: ${result.name}`);
}

/** Workflow step: print the description and succeed. */
export function describe(
  _stepFns: any,
  opts: Opts,
  reportFn: (opts: Opts) => DescribeResult = describeReport,
): Opts {
  const result = reportFn(opts);
  printReport(result);
  return okAlias({ ...opts, "describe/result": result });
}

File: create-py-bigconfig-package/templates/src/pkg/describe.py
"""Simple package description. Pure report + workflow step."""
from __future__ import annotations

from typing import Any, Callable, TypedDict

from big_config.core import Opts

from .interop import ok_alias, params_of, profile_of, read_bc_pars


class DescribeResult(TypedDict):
    profile: str
    package: str
    name: str


def _display(value: Any, fallback: str = "(unset)") -> str:
    if isinstance(value, str) and value.strip() != "":
        return value
    if value is None:
        return fallback
    return str(value)


def describe_report(opts: Opts, env: dict[str, str | None] | None = None) -> DescribeResult:
    """Pure: build a concise description of the active profile."""
    merged = read_bc_pars(opts, env)
    params = params_of(merged)
    return {
        "profile": _display(profile_of(merged), "default"),
        "package": _display(params.get("package")),
        "name": _display(params.get("name")),
    }


def print_report(result: DescribeResult) -> None:
    print(f"Package: {result['package']}")
    print(f"Profile: {result['profile']}")
    print(f"Name: {result['name']}")


def describe(_step_fns: Any, opts: Opts, report_fn: Callable[[Opts], DescribeResult] = describe_report) -> Opts:
    """Workflow step: print the description and succeed."""
    result = report_fn(opts)
    print_report(result)
    return ok_alias({**opts, "describe/result": result})
